From 3fba6f82ab1199810396411070928d47a89b2cab Mon Sep 17 00:00:00 2001
From: Pantelis Antoniou <panto@antoniou-consulting.com>
Date: Mon, 20 May 2013 16:01:07 +0300
Subject: [PATCH 10/10] tilcdc: Only support Audio on 50 & 60 Hz modes

Apparently anything other fails to work (at least for me).

Signed-off-by: Pantelis Antoniou <panto@antoniou-consulting.com>
---
 drivers/gpu/drm/tilcdc/tilcdc_crtc.c | 20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/tilcdc/tilcdc_crtc.c b/drivers/gpu/drm/tilcdc/tilcdc_crtc.c
index c86da03..779f31f 100644
--- a/drivers/gpu/drm/tilcdc/tilcdc_crtc.c
+++ b/drivers/gpu/drm/tilcdc/tilcdc_crtc.c
@@ -310,7 +310,7 @@ static int tilcdc_crtc_mode_set(struct drm_crtc *crtc,
 	    reg |= (((mode->hdisplay >> 4) - 1) & 0x40) >> 3;
 
 	tilcdc_write(dev, LCDC_RASTER_TIMING_0_REG, reg);
-	
+
 	/* only the vertical sync width maps 0 as 1 so only subtract 1 from vsw */
 	reg = ((mode->vdisplay - 1) & 0x3ff) |
 	    ((vbp & 0xff) << 24) |
@@ -440,7 +440,7 @@ int tilcdc_crtc_mode_valid(struct drm_crtc *crtc, struct drm_display_mode *mode,
   	struct tilcdc_drm_private *priv = crtc->dev->dev_private;
 	unsigned int bandwidth;
 	uint32_t hbp, hfp, hsw, vbp, vfp, vsw;
-	int has_audio, is_cea_mode;
+	int has_audio, is_cea_mode, can_output_audio, refresh;
 
 	int rb;
 
@@ -459,16 +459,24 @@ int tilcdc_crtc_mode_valid(struct drm_crtc *crtc, struct drm_display_mode *mode,
 	/* set if there's audio capability */
 	has_audio = edid && drm_detect_monitor_audio(edid);
 
+	/* only 50 & 60Hz modes reliably support audio */
+	refresh = drm_mode_vrefresh(mode);
+
 	/* set if it's a cea mode */
 	is_cea_mode = drm_match_cea_mode(mode) > 0;
 
-	DBG("mode %dx%d@%d pixel-clock %d audio %s cea %s",
-		mode->hdisplay, mode->vdisplay, drm_mode_vrefresh(mode),
+	/* set if we can output audio */
+	can_output_audio = edid && has_audio && is_cea_mode &&
+		(refresh == 50 || refresh == 60);
+
+	DBG("mode %dx%d@%d pixel-clock %d audio %s cea %s can_output %s",
+		mode->hdisplay, mode->vdisplay,refresh,
 		mode->clock,
 		has_audio ? "true" : "false",
-		is_cea_mode ? "true" : "false");
+		is_cea_mode ? "true" : "false",
+		can_output_audio ? "true" : "false" );
 
-	if (edid && audio && has_audio && !is_cea_mode) {
+	if (edid && has_audio && !can_output_audio) {
 		DBG("Pruning mode : Does not support audio");
 		return MODE_BAD;
 	}
-- 
1.8.1.4

